cgp/ea/mu_plus_lambda.py

```diff
@@ -137,13 +137,34 @@ def step(
         combined = self._compute_fitness(combined, objective)
         combined = self._sort(combined)
 
-        n_total = self.n_offsprings + pop.n_parents
-        k_local_search = n_total if self.k_local_search is None else self.k_local_search
-        for idx in range(k_local_search):
-            self.local_search(combined[idx])
+        if self.local_search is not None:
+            assert isinstance(pop.champion.fitness, float)
+            prev_max_fitness: float = pop.champion.fitness
 
-        combined = self._compute_fitness(combined, objective)
-        combined = self._sort(combined)
+            combined_copy = [ind.copy() for ind in combined]
```

Comment:
why do we need the combined copy here and below?

Reply:
we perform the local search on a copy of the population since we might want to revert back to the original parameters if the fitness has decreased due to local search. to revert we need to keep both the original and a copy. l154ff then chooses which one to keep.

Reply:
I see. But if we do that, can the assert statement above ever be triggered? It seems that this manoeuvre should prevent the assert statement discussed above from being triggered altogether, shouldn't it?

Reply:
it should never be triggered, yes :+1:

Reply:
So, we can remove both the assert statement and the copy ?

Reply:
nope, copy is necessary. the copy is the thing that makes sure only fitness improvements are accepted. regarding the assert, see above.

Reply:
Okay, understood.